{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eecc2468-1d01-4cf7-b2eb-2e74b9558f16",
   "metadata": {},
   "outputs": [],
   "source": [
    "# D. coronatus HiC PCA\n",
    "# run the Arima pipeline\n",
    "# modified from the original at Arima’s GitHub\n",
    "# https://github.com/ArimaGenomics/mapping_pipeline\n",
    "# the output of this pipeline is a BAM file: Dco_hiC_rep1.bam\n",
    "sbatch JOB_2024-11-18_003_Dco-HiC_Arima_pipe.sh\n",
    "\n",
    "# The main result of the pipe is a bam file with mappings for both ends of the reads\n",
    "# in ./deduplicated_files\n",
    "# Use bam2pairs from pairix to convert into .pairs file\n",
    "# Then use juicer_tools to convert .pairs into .hic\n",
    "# nxDipCoro1_1.curated_primary.chrom.sizes contains 3 lines and 2 columns:\n",
    "# First column being the chromosome name, second the size in bps\n",
    "conda create -n pairix-env -c bioconda pairix\n",
    "conda activate pairix-env\n",
    "bam2pairs -l Dco_hiC_rep1.bam Dco_hiC_rep1\n",
    "java -Xmx64g -jar juicer_tools.jar pre Dco_hiC_rep1.bsorted.pairs.gz Dco_hiC_rep1.bsorted.hic ./nxDipCoro1_1.curated_primary.chrom.sizes\n",
    "conda deactivate\n",
    "\n",
    "# Try principal component analysis with hicexplorer\n",
    "#\n",
    "# convert into .cool, resolution 100,000 bps\n",
    "conda create -n hicexplorer-env -c bioconda hicexplorer\n",
    "conda activate hicexplorer-env\n",
    "hicConvertFormat --matrices Dco_hiC_rep1.bsorted.hic --inputFormat hic -r 100000 --outFileName Dco_hiC_rep1.bsorted.cool --outputFormat cool\n",
    "\n",
    "# convert to h5\n",
    "hicConvertFormat --matrices Dco_hiC_rep1.bsorted_100000.cool --inputFormat cool -r 100000 --outFileName Dco_hiC_rep1.bsorted_100000.h5 --outputFormat h5\n",
    "\n",
    "# do a diagnostic before matrix correction\n",
    "hicCorrectMatrix diagnostic_plot -m Dco_hiC_rep1.bsorted_100000.cool -o Dco_hiC_rep1.bsorted_100000.diag.png\n",
    "\n",
    "# attempt to correct the matrix, using ICE\n",
    "hicCorrectMatrix correct --matrix Dco_hiC_rep1.bsorted_100000.h5 --filterThreshold -3 3 --correctionMethod ICE --outFileName Dco_hiC_rep1.bsorted_100000.correctedICE.h5\n",
    "\n",
    "# do the PCA\n",
    "# 5 PCs\n",
    "hicPCA --matrix Dco_hiC_rep1.bsorted_100000.correctedICE.h5 --outputFileName Dco_hiC_rep1.bsorted_100000.correctedICE.pca1.bedgraph Dco_hiC_rep1.bsorted_100000.correctedICE.pca2.bedgraph Dco_hiC_rep1.bsorted_100000.correctedICE.pca3.bedgraph Dco_hiC_rep1.bsorted_100000.correctedICE.pca4.bedgraph Dco_hiC_rep1.bsorted_100000.correctedICE.pca5.bedgraph --whichEigenvectors 1 2 3 4 5 --format bedgraph --ignoreMaskedBins\n",
    "\n",
    "# The PC that corresponds to long-range inter-/intra-chromosome interactions is PC1\n",
    "# Take the bedgraph file and convert it into the format similar to all the traits being graphed, ie:\n",
    "# scaffold_position    PC1\n",
    "# 50000                0\n",
    "# 150000               0.01\n",
    "# ...etc.-\n",
    "# PCA_bedgraph_converter.py takes the average of all values within 100,000-bp windows and plots the average in the middle of the interval (ie. %100,000 = 50,000)\n",
    "# nxDipCoro1_1.curated_primary.fa is the D. coronatus genome fasta\n",
    "# \n",
    "# The command below should give the files \"Dcor.SUPER_1.PC1.window100000.increment100000.txt\" and \"Dcor.SUPER_2.PC1.window100000.increment100000.txt\"\n",
    "python3 PCA_bedgraph_converter.py -f nxDipCoro1_1.curated_primary.fa -b Dco_hiC_rep1.bsorted_100000.correctedICE.pca1.bedgraph -x scaffold_position -y PC1 -p Dcor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56d99d49-d256-4b3b-b45d-7767c5995fa0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# D. pachys PoreC PCA\n",
    "# First, PoreC homozygous region matrix transformation\n",
    "# This is necessary because a homozygous region exists between DpaA and DpaB. The homozygous region receives 2x the number of reads mapping.\n",
    "# In order to do PCA, we randomly assigned reads mapping to this region to either DpaA or DpaB.\n",
    "#\n",
    "# Input file is \"NlaIII_run01_Dpa-canu-het00c-YaHS-v202304-hethomosep_unphased.unsorted.pairs\"\n",
    "# which contains reads assigned to DpaA heterzygous region, DpaB heterozygous region, and the homozygous region\n",
    "#\n",
    "# This Python script randomly assign reads mapping to the homozygous region to either DpaA (scaffold_1) or DpaB (scaffold_2)\n",
    "# The output of this script should be \"Dpa-canu-het00c-YaHS-v202304.pairs\"\n",
    "python3 Dpa-canu-het00c-YaHS-v202304-hethomosep_PoreC_correction.py\n",
    "\n",
    "# use pairtools (version 1.0.2) to sort the .pairs file\n",
    "# pairtools: https://github.com/open2c/pairtools\n",
    "conda create -n pairtools-env -c conda-forge -c bioconda pairtools\n",
    "conda activate pairtools-env\n",
    "pairtools sort Dpa-canu-het00c-YaHS-v202304.pairs\n",
    "\n",
    "# use juicer_tools.jar pre to generate a hic file from the .sorted.pairs file\n",
    "# juicer_tools.jar can be found at: https://github.com/aidenlab/JuicerTools\n",
    "# chrom.sizes is just a tab-delimited chromosome_name, chromosome_size_in_bps\n",
    "# (chromosomes separated by return character)\n",
    "java -Xmx2g -jar juicer_tools.jar pre Dpa-canu-het00c-YaHS-v202304.sorted.pairs Dpa-canu-het00c-YaHS-v202304.hic ./chrom.sizes\n",
    "\n",
    "# try principal component analysis with hicexplorer (version 3.7.2)\n",
    "# convert into cool, resolution 100,000 bps\n",
    "conda deactivate\n",
    "conda activate hicexplorer-env\n",
    "hicConvertFormat --matrices Dpa-canu-het00c-YaHS-v202304.hic --inputFormat hic -r 100000 --outFileName Dpa-canu-het00c-YaHS-v202304.cool --outputFormat cool\n",
    "\n",
    "# convert to h5\n",
    "hicConvertFormat --matrices Dpa-canu-het00c-YaHS-v202304_100000.cool --inputFormat cool -r 100000 --outFileName Dpa-canu-het00c-YaHS-v202304_100000.h5 --outputFormat h5\n",
    "\n",
    "# do a diagnostic before matrix correction\n",
    "hicCorrectMatrix diagnostic_plot -m Dpa-canu-het00c-YaHS-v202304_100000.cool -o Dpa-canu-het00c-YaHS-v202304_100000.diag.png\n",
    "\n",
    "# attempt to correct the matrix, using the ICE algorithm\n",
    "hicCorrectMatrix correct --matrix Dpa-canu-het00c-YaHS-v202304_100000.h5 --filterThreshold -3.5 3.5 --correctionMethod ICE --outFileName Dpa-canu-het00c-YaHS-v202304_100000.correctedICE.h5\n",
    "\n",
    "# attempt to do PCA, with 5 PCs\n",
    "# output as .bedgraph file\n",
    "hicPCA --matrix Dpa-canu-het00c-YaHS-v202304_100000.correctedICE.h5 --outputFileName Dpa-canu-het00c-YaHS-v202304_100000.correctedICE.pca1.begraph Dpa-canu-het00c-YaHS-v202304_100000.correctedICE.pca2.bedgraph Dpa-canu-het00c-YaHS-v202304_100000.correctedICE.pca3.bedgraph Dpa-canu-het00c-YaHS-v202304_100000.correctedICE.pca4.bedgraph Dpa-canu-het00c-YaHS-v202304_100000.correctedICE.pca5.bedgraph --numberOfEigenvectors 5 --format bedgraph –ignoreMaskedBins\n",
    "conda deactivate\n",
    "\n",
    "# Take the bedgraph file and convert it into the format similar to all the traits being graphed, ie:\n",
    "# scaffold_position    PC3\n",
    "# 50000                0\n",
    "# 150000               0.01\n",
    "# ...etc.-\n",
    "# PCA_bedgraph_converter.py takes the average of all values within 100,000-bp windows and plots the average in the middle of the interval (ie. %100,000 = 50,000)\n",
    "# Dpa-canu-het00c-YaHS-v202304.fasta is the genome file\n",
    "# \n",
    "# The command below should give the files \"Dpac.scaffold_1.PC3.window100000.increment100000.txt\" and \"Dpac.scaffold_2.PC3.window100000.increment100000.txt\"\n",
    "python3 PCA_bedgraph_converter.py -p Dpac -f Dpa-canu-het00c-YaHS-v202304.fasta -b Dpa-canu-het00c-YaHS-v202304_100000.correctedICE.pca3.bedgraph -w 100000 -i 100000 -x scaffold_position -y PC3"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
