# Supplemental Figure 2: *D. coronatus* telomeric sequences are added to various points along the most distal subtelomeric repeat unit.

Run the ```Diploscapter_coronatus_telomere_gggenomes_visualization_2025-12-16.py``` code. This code requires the ```*.sorted.coords```, ```*.telOccupancyDrop.txt```, and ```*.telomereScore20.fasta``` files generated for **Fig. 2G**.
- ```*.sorted.coords``` contains the coordinates of subtelomeric repeat units, sorted by their positions on the sequencing reads.
- ```*.telOccupancyDrop.txt``` contains the position of the most disstal subtelomeric repeat unit, and how it compares with the position where there is a steep drop in the occupancy of TAAGG[G/C] repeats on the sequencing reads.
- ```*.telomereScore20.fasta``` contains the likely telomeric reads, in FASTA format.


Running the script:
```bash
python3 Diploscapter_coronatus_telomere_gggenomes_visualization_2025-12-16.py
```

**Supplemental Fig. 2** was generated by pasting the text output ```Dco.zoomed_in_bases.no_gaps.txt``` into Adobe Illustrator for formatting and annotating. Schematics of the *D. coronatus* telomeric reads (not used in the publication) can be generated using the gggenomes package (https://thackl.github.io/gggenomes/).

<img width="4200" height="1487" alt="image" src="https://github.com/user-attachments/assets/e06a91f0-faf6-4e1d-ae01-56ab517046f4" />
